### PR TITLE
Implement memoize

### LINKED ISSUE #1: implement memoize

### PR DESCRIPTION
**Pull request by greg-w101**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -24,29 +21,23 @@ jobs:
...
         run: npm audit signatures
-        env:
-          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
       - name: Compile
         run: npm run ci:compile
       - name: Check Code Style
         run: npm run ci:style
       - name: Perform Linting
         run: npm run ci:lint
-      - name: Run Tests
```
Comment (carlansley) [inline .github/workflows/ci.yml]: we should be running tests

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -58,19 +49,13 @@ jobs:
...
```
Comment (carlansley) [inline .github/workflows/ci.yml]: we should be running tests

File: LICENSE.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+
+The MIT License
+
+Copyright (c) 2021-2023 Check Digit, LLC
```
Comment (carlansley) [inline LICENSE.txt]: this is a new repo, should just be 2023

File: README.md [hunk 4]
```diff
@@ -1,36 +1,25 @@
-# Check Digit < Insert name > Library
+# Memoize
 
-The Check Digit < Insert Name > library is a library for Check Digit services to deal with [ short description about what the library does]
+Copyright (c) 2021-2023 [Check Digit, LLC](https://checkdigit.com)
```
Comment (carlansley) [inline README.md]: 2023

File: README.md [hunk 5]
```diff
@@ -1,36 +1,25 @@
...
 
-Features:
+Memoize is a light-weight utility library used by Check Digit services for memoization purposes.
```
Comment (carlansley) [inline README.md]: I think we should add an explantation that its designed to memoize promises, and evict from the cache anything that rejects.

File: package.json [hunk 6]
```diff
@@ -1,52 +1,75 @@
 {
-  "name": "@checkdigit/library-template",
-  "version": "0.0.1",
-  "description": "A short description of the intended use of the library",
-  "typings": "./dist/index.d.ts",
-  "main": "./dist/index.js",
+  "name": "@checkdigit/memoize",
+  "version": "1.0.0",
+  "description": "Memoize is a light-weight utility library used by Check Digit services for memoization purposes",
```
Comment (carlansley) [inline package.json]: ...for memoizing promises with cache eviction on reject

File: src/index.ts [hunk 7]
```diff
@@ -1,3 +1,9 @@
 // index.ts
 
-export { exampleFunction } from './example-directory';
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
Comment (carlansley) [inline src/index.ts]: 2023

File: src/memoize.spec.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
+// memoize.spec.ts
+
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
Comment (carlansley) [inline src/memoize.spec.ts]: 2023

File: src/memoize.spec.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import { strict as assert } from 'node:assert';
+import memoize from './memoize';
```
Comment (carlansley) [inline src/memoize.spec.ts]: I think it should import memoize from `index.ts`, for complete code coverage

File: src/memoize.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
+// memoize.ts
+
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
Comment (carlansley) [inline src/memoize.ts]: 2023

File: src/memoize.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import type { URL } from 'node:url';
+
+// arguments must be stringify-able
+export type Argument = string | number | boolean | null | Date | URL | Argument[] | { [key: string]: Argument };
+
+/**
+ * Memoize an async function.  The promise is cached, not the value.  This guarantees the underlying function is
```
Comment (carlansley) [inline src/memoize.ts:15]: a version of this comment probably should be in the README

**Review by carlansley — CHANGES_REQUESTED** [review]

**Review by lukebrdn — DISMISSED** [review]
lgtm

File: README.md [hunk 12]
```diff
@@ -1,36 +1,28 @@
-# Check Digit < Insert name > Library
+# Memoize
 
-The Check Digit < Insert Name > library is a library for Check Digit services to deal with [ short description about what the library does]
+Copyright (c) 2023 [Check Digit, LLC](https://checkdigit.com)
 
-Features:
+Memoize is a small async function used by Check Digit services for memoizing promises with cache eviction on reject.
```
Comment (carlansley) [inline README.md]: remove the "used by Check Digit services"

File: README.md [hunk 13]
```diff
@@ -1,36 +1,28 @@
...
 ```
-import * as insertName from '@checkdigit/< Insert name >';
+$ npm install @checkdigit/memoize
+```
 
-const someVariable = insertName();
+### Use
 
 ```
+import { memoize } from '@checkdigit/memoize';
```
Comment (carlansley) [inline README.md]: I don't think this code would work, there's only a default import.   believe should be `import memoize from '@checkdigit/memoize';`

**Review by carlansley — CHANGES_REQUESTED** [review]

**Comment by greg-w101** [comment]
> lgtm

Luke you left a comment instead of approving haha

**Review by lukebrdn — APPROVED** [review]
lgtm

**Review by carlansley — APPROVED** [review]

**Pull request merged**